### DISCUSSION TITLE
Subgraphs

### DISCUSSION THREAD

**Discussion by jejacks0n** [alelievr/NodeGraphProcessor#141] [General]
No issues with the project, it's awesome, and I wanted to say thanks. I've been playing around for several hours today, and this work really have saved me a ton of time.

I'm looking for subgraphs, and am willing to help contribute to that. How can I help? I'm still trying to wrap my head around the project, but I'm considering approaching it from a custom contextual menu item, and then building an "output" node similar to how shadergraph is setup. "Subgraphs" would have some validation around requiring that node. The rest is mostly fuzzy, but doesn't seem that hard to either call into existing logic for processing the subgraph "node", or add those concepts (but I suspect they already exist more or less).

Obviously you're probably much better suited to build that correctly, but let me know if and how I can help.

Thanks again! Really great library -- this is what I was hoping for from unity, but that's a bit messier than what you provide here. Good work. =)

**Comment by alelievr** [comment]
Hello :)

Thanks for wanting to contribute to this package! Subgraphs have been on the roadmap for quite a long time but I never did it.
Now that the serialization issue has been fixed (in the latest big update) I think nothing prevents us from adding subgraphs :)

Your idea of subgraphs is pretty similar to what I was thinking, which is:
An asset derived from the BaseGraph class and stored as an asset on the disk, then we have a node that references this asset inside a graph, exposes the graph parameters as inputs/output in the node (parameters can be configured in both input and output). I don't think the subgraph needs an output node (because parameters can do that) and it would limit the system to a monolithic main output which may not be what the user wants when creating a subgraph.

We can then inject the subgraphs directly into the node creation menu (it shouldn't be hard to do).

After that I see two possible implementations in term of processing:
- The first is to flatten the graph and subgraphs before processing all the nodes (that way we can also solve the cyclic subgraph problem). This is similar to what unreal blueprint does and it allows to optimize the graph processing in certain cases like this one:
![image](https://user-images.githubusercontent.com/6877923/108003434-1a112780-6ff3-11eb-9b2b-d39833cf31c8.png)
The blue box is a subgraph, and as you can see the upper branch is not required because its output is not connected.

- The second approach is much simpler and simply consists of calling the Process of the subgraph asset contained in the subgraph node when the node is processed. This is obviously much simpler to do but you can't optimize anything like this and it becomes quite tricky to handle the cyclic subgraph case.

What do you think about it?

PS: I'll convert this issue into a discussion as we'll continue to talk about this :)
